### PR TITLE
8281267: VM HeapDumper dumps array classes several times

### PR DESCRIPTION
**Pull request by alexmenkov**
ClassLoaderDataGraph::classes_do description says:
// Walking classes through the ClassLoaderDataGraph include array classes.
So do_load_class callback should not dump arrays for the classes and dumper doesn't need to call Universe::basic_type_classes_do (array classes for primitive types are also reported by ClassLoaderDataGraph::classes_do)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8281267](https://bugs.openjdk.java.net/browse/JDK-8281267): VM HeapDumper dumps array classes several times

### Reviewers
 * [Chris Plummer](https://openjdk.java.net/census#cjplummer) (@plummercj - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7384/head:pull/7384` \
`$ git checkout pull/7384`

Update a local copy of the PR: \
`$ git checkout pull/7384` \
`$ git pull https://git.openjdk.java.net/jdk pull/7384/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7384`

View PR using the GUI difftool: \
`$ git pr show -t 7384`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7384.diff">https://git.openjdk.java.net/jdk/pull/7384.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/serviceability/HeapDump/DuplicateArrayClassesTest.java [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+ * @run driver DuplicateArrayClassesTest
+ */
+
+class DuplicateArrayClassesTarg extends LingeredApp {
+    public static void main(String[] args) {
+        // Initialize some array classes (for primitive type and object type).
+        int[][][] intArray = new int[0][][];
+        String[][][] strArray = new String[0][][];
+        LingeredApp.main(args);
+        System.out.println("" + intArray + strArray);   // to be sure the classes are not unloaded
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/HeapDump/DuplicateArrayClassesTest.java:58]: Use  Reference.reachabilityFence(<array>);

File: test/hotspot/jtreg/serviceability/HeapDump/DuplicateArrayClassesTest.java [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+            //jcmd <pid> GC.heap_dump <file_path>
+            JDKToolLauncher launcher = JDKToolLauncher
+                    .createUsingTestJDK("jcmd")
+                    .addToolArg(Long.toString(theApp.getPid()))
+                    .addToolArg("GC.heap_dump")
+                    .addToolArg(dumpFile.getAbsolutePath());
+            Process p = ProcessTools.startProcess("jcmd", new ProcessBuilder(launcher.getCommand()));
+            // If something goes wrong with heap dumping most likely we'll get crash of the target VM.
+            while (!p.waitFor(5, TimeUnit.SECONDS)) {
```
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/HeapDump/DuplicateArrayClassesTest.java:86]: 5 seconds seems kind of short. Doesn't allow for occasional network hiccups.

File: src/hotspot/share/services/heapDumper.cpp [hunk 3]
```diff
@@ -2291,17 +2283,19 @@ void VM_HeapDumper::work(uint worker_id) {
...
     }
-    Universe::basic_type_classes_do(&do_load_class);
 
     // write HPROF_FRAME and HPROF_TRACE records
     // this must be called after _klass_map is built when iterating the classes above.
     dump_stack_traces();
 
     // Writes HPROF_GC_CLASS_DUMP records
+    // For array classes we need signers and protection domain from their bottom classes
+    // so do_class_dump skips array classes and dumps array classes with corresponding instance class.
```
Comment (plummercj) [inline src/hotspot/share/services/heapDumper.cpp:2293]: I'm having trouble parsing this comment. Can you clarify?

Comment (alexmenkov) [reply src/hotspot/share/services/heapDumper.cpp:2293]: HPROF_GC_CLASS_DUMP records contain references to signers and protection domain objects.
But this is properties of InstanceKlass, for array classes the properties from their element classes are written (or NULL for arrays of primitive types).
do_class_dump callback skips array classes:
  if (k->is_instance_klass()) {
    DumperSupport::dump_class_and_array_classes(writer(), k);
  }

dump_class_and_array_classes dumps the class and then iterate over array classes (k->array_klass_or_null()) and writes records for them (using signer and protection domain from the instance class specified).

This approach also skips primitive type arrays (they don't have corresponding InstanceKlass) - they are dumped separately.

I added the comment to explain why do_class_dump goes this tricky way.
Do you have any suggestions how to rephrase it to make clearer?

File: test/hotspot/jtreg/serviceability/HeapDump/DuplicateArrayClassesTest.java (hunk 1 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/HeapDump/DuplicateArrayClassesTest.java:58]: Fixed

File: test/hotspot/jtreg/serviceability/HeapDump/DuplicateArrayClassesTest.java (hunk 2 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/HeapDump/DuplicateArrayClassesTest.java:86]: Note that this is a cycle. It ensures target VM (theApp) is not terminated while jcmd (p) processes the command.
I had some experience in the past when jcmd didn't exit if target VM crashes.

File: test/hotspot/jtreg/serviceability/HeapDump/DuplicateArrayClassesTest.java [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+                log("HPROF_LOAD_CLASS records (" + duplicateLoadClassIDs.size() + "):");
+                duplicateLoadClassIDs.forEach(id -> log("  - id = " + id + ": " + classId2Name.get(id)));
+                log("HPROF_GC_CLASS_DUMP records (" + duplicateDumpClassIDs.size() + "):");
+                duplicateDumpClassIDs.forEach(id -> log("  - id = " + id + ": " + classId2Name.get(id)));
+                throw new RuntimeException("duplicates detected");
+            }
+        }
+    }
+
+    // Reads the whole HPROF_GC_CLASS_DUMP record, returns closs ID.
```
[the remark this answers is not in the document]
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/HeapDump/DuplicateArrayClassesTest.java:218]: Fixed.

File: src/hotspot/share/services/heapDumper.cpp (hunk 3 above)
Comment (plummercj) [reply src/hotspot/share/services/heapDumper.cpp:2293]: Sorry about the delay. I didn't see your response when it came in.

It seems like this comment belongs with do_class_dump(). Putting it here is kind of confusing. Also, what is meant by "bottom class"? Do you mean the element type of the array?

Comment (alexmenkov) [reply src/hotspot/share/services/heapDumper.cpp:2293]: This comment explains why we call ClassLoaderDataGraph::classes_do(do_class_dump) and then Universe::basic_type_classes_do(&do_basic_type_array_class_dump).

Bottom klass is 1-dimensional type.
For example for [[LObject; element type is [lObject;, and bottom klass is LObject;
Looked at the code and see that ObjAyyaKlass actually contains link to it - see ObjArrayKlass::bottom_klass(), so I can update do_class_dump to use it.

Comment (alexmenkov) [reply src/hotspot/share/services/heapDumper.cpp]: I updated the fix, fixed do_class_dump and removed this comment.
This also removes code duplication in dump_class_and_array_classes and dump_basic_type_array_class

**Review by plummercj — APPROVED** [review]
The changes look good. Have you verified that after your most recent changes you get the same sized hprof file as with your initial changes?

**Comment by alexmenkov** [comment]
> The changes look good. Have you verified that after your most recent changes you get the same sized hprof file as with your initial changes?

Thank you for the review.
I verified number of classes (HPROF_LOAD_CLASS and HPROF_GC_CLASS_DUMP records) are the same with initial change and with the most recent (and the numbers of HPROF_GC_CLASS_DUMP records are the same as without the fix).
Size of the hprof file changes slightly from run to run (+-1KB for 2.5MB hprof file), so we can't rely on it

File: src/hotspot/share/services/heapDumper.cpp [hunk 5]
```diff
@@ -2027,52 +1987,41 @@ void DumperSupport::end_of_dump(AbstractDumpWriter* writer) {
...
-    // class ID
-    Klass* klass = k;
-    writer()->write_classID(klass);
+  // class serial number is just a number
+  writer()->write_u4(++class_serial_num);
 
-    // add the Klass* and class serial number pair
-    dumper()->add_class_serial_number(klass, class_serial_num);
+  // class ID
+  Klass* klass = k;
```
Comment (coleenp) [inline src/hotspot/share/services/heapDumper.cpp:2003]: I don't think it's necessary to introduce the variable 'klass' since it's not a loop anymore.

File: src/hotspot/share/services/heapDumper.cpp [hunk 6]
```diff
@@ -2302,7 +2250,6 @@ void VM_HeapDumper::work(uint worker_id) {
       LockedClassesDo locked_dump_class(&do_class_dump);
       ClassLoaderDataGraph::classes_do(&locked_dump_class);
     }
-    Universe::basic_type_classes_do(&do_basic_type_array_class_dump);
```
Comment (coleenp) [inline src/hotspot/share/services/heapDumper.cpp:2305]: If this is the last use of this function, you could remove this variant of this function from Universe.

**Review by coleenp — APPROVED** [review]
This looks like a nice cleanup!  Thank you for the test and verification that the format is the same.  A couple of minor comments.

File: src/hotspot/share/services/heapDumper.cpp (hunk 6 above)
Comment (alexmenkov) [reply src/hotspot/share/services/heapDumper.cpp:2305]: It's still used in JvmtiGetLoadedClasses::getClassLoaderClasses()

File: src/hotspot/share/services/heapDumper.cpp (hunk 5 above)
Comment (alexmenkov) [reply src/hotspot/share/services/heapDumper.cpp]: Fixed

File: src/hotspot/share/services/heapDumper.cpp (hunk 6 above)
Comment (coleenp) [reply src/hotspot/share/services/heapDumper.cpp:2305]: > void Universe::basic_type_classes_do(void f(Klass*)) {
This one is unused.  The other one that takes the Closure is the one that getClassLoaderClasses() calls.

Comment (alexmenkov) [reply src/hotspot/share/services/heapDumper.cpp:2305]: I thought we always have 2 version (for function pointer and closure object), Universe::metaspace_pointers_do has only version for closures.
Will remove unused one.

**Review by coleenp — APPROVED** [review]
Thanks!
